Charles/cli port command

charlesHetterich opened the pull request:
Thank you for your Pull Request! 🙏

✄ -----------------------------------------------------------------------------

# Description

*A concise description of what your PR is doing, and what potential issue it is solving. Use [Github semantic
linking](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
to link the PR to an issue that must be closed once this is merged.*

# Checklist

* [ ] My PR includes a detailed description as outlined in the "Description" and its two subsections above.
* [ ] I have made corresponding changes to the documentation (if applicable)
* [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)

Remove the "Checklist" section once all have been checked. Thank you for your contribution!

charlesHetterich commented:
You may validate & observe behavior against a variety of SCC Testing repos in a fresh directory by running the scripts
`clean-clone.sh`
```
# Clean existing repos
for dir in */; do
    if [ -d "${dir}/.git" ]; then
        echo "Removing git repository: $dir"
        rm -rf "$dir"
    fi
done

git clone https://github.com/axelarnetwork/interchain-token-service.git      # interchain-token-service [Axelar]
git clone https://github.com/redstone-finance/redstone-oracles-monorepo.git  # redstone-oracles-monorepo
git clone https://github.com/fireblocks/fireblocks-smart-contracts.git       # fireblocks-smart-contracts
git clone https://github.com/smartcontractkit/chainlink-evm.git              # chainlink-evm
git clone https://github.com/aave/ccip.git                                   # Aave CCIP
```

`test-all.sh`
```
test_port() {
    pnpm dlx tsx <YOUR-PATH-TO>/hardhat-polkadot/packages/hardhat-polkadot/src/cli port "$1" "$2"
}

test_port "./interchain-token-service" "$1"
test_port "./redstone-oracles-monorepo/packages/eth-contracts" "$1"
test_port "./fireblocks-smart-contracts" "$1"
test_port "./chainlink-evm/contracts" "$1"
test_port "./ccip/contracts" "$1"
```

And just replace `<YOUR-PATH-TO>` appropriately.

charlesHetterich commented:
Just FYI I noticed CI being inconsistent with failing/passing— has this been an issue or is this new?
- for Unit tests it looks like it sometimes fails to connect to NPM registry
- for E2E tests I'm not sure because the process exits without an error message

UtkarshBhardwaj007 commented on packages/hardhat-polkadot/src/cli/port-project.ts:
```diff
@@ -0,0 +1,123 @@
+    // Update `hardhat` version if needed
+    const HHLocation = pkg.dependencies?.hardhat ? "dependencies" : "devDependencies"
+    const currentHHVersion = pkg[HHLocation]?.hardhat?.match(/(\d+)\.(\d+)\.(\d+)/)?.[0] || "0.0.0"
+    if (currentHHVersion < requiredHHVersion) {
```
AFAIK this plugin also doesn't work with hardhat  >= 3.0.0, worth adding a check for that.

charlesHetterich replied:
I am defaulting to throw an error like `Unsupported hardhat version 3.0.4. Please manually install a compatible version (^2.26.0)` for this case

Let me know if we'd prefer to automatically downgrading their HH version instead

UtkarshBhardwaj007 commented on packages/hardhat-polkadot/src/cli/port-project.ts:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    // Add `@parity/hardhat-polkadot` dev-dependency
+    pkg.devDependencies["@parity/hardhat-polkadot"] = `^${polkadotHHVersion}`
```
Do you need to check for NPE here, if devDependencies doesn't exist?

UtkarshBhardwaj007 commented on packages/hardhat-polkadot/src/cli/port-project.ts:
```diff
@@ -0,0 +1,123 @@
+export async function updatePackageJSON(projectPath: string): Promise<[string, string, boolean]> {
+    // Read `package.json` object & formatting metadata
+    const pkgPath = path.join(projectPath, "package.json")
+    const raw = fs.readFileSync(pkgPath, "utf8")
```
Should check if package.json exits. If not, give an appropriate error log.

charlesHetterich replied:
I added a comment explaining how we handle errors here

charlesHetterich commented on packages/hardhat-polkadot/src/cli/port-project.ts:
```diff
@@ -0,0 +1,123 @@
+    const indent = detectIndent(raw)
+
+    // Fetch latest `@parity/hardhat-polkadot` module metadata from npm registry
+    const m = (await (await fetch(`https://registry.npmjs.org/${MODULE}/latest`)).json()) as {
```
Similar to my comment above— when `fetch` itself fails, we want the natural error to bubble up so that it gives the best context as to what the error really is.

However I see it is possible for fetch *not to throw* here, but produce an invalid `m`, So I'm adding the following check to handle that error case
```
    if (!m.version || !m.peerDependencies) {
        throw new Error(`Something went wrong fetching ${MODULE} metadata from npm registry`)
    }
```

I will also add a comment which explains our error handling method

charlesHetterich commented on packages/hardhat-polkadot/src/cli/file-utils.ts:
```diff
@@ -0,0 +1,139 @@
+ *
+ * @returns a tuple of [gitIgnorePath, originalContent, newContent]
+ */
+export async function addOrMergeGitIgnore(projectRoot: string): Promise<[string, string, string]> {
```
I looked into this one, but this package is for file filtering (using the same spec as gitignore). Its not actually useful for updating the contents of an existing .gitignore file. Afaik, there is no well-maintained package which does this (i.e. cleanly merge partially redundant content into an existing .gitignore with minimal diffs)

charlesHetterich commented on packages/hardhat-polkadot/src/cli/file-utils.ts:
```diff
@@ -0,0 +1,139 @@
+
+/**
+ * Prints diff between `before` and `after` versions of a file
+ * NOTE: assumes `git` command is available
```
Resolved this one by generating diffs with the `diff` package instead of using git

charlesHetterich commented:
I've addressed all comments besides for testing, which I plan on addressing in the morning.

My current plan would be an E2E style of testing by creating mock projects in
```
tests/fixture-projects/
    test-port-cmd/
        project-scenario-1/
            package.json
            expected_package.json
            .gitignore
            ...
        project-scenario-2/...
        ...
```
and updating the appropriate testing script to actually run & validate the output of the `port` command against each mock project.

Utkarsh mentioned unit testing for individual utility functions— which in general I agree with, but there's currently no infrastructure for this and I don't see it being done elsewhere, so I wasn't planning on that.

Please let me know if you prefer I change my approach or do unit testing, thanks!

charlesHetterich commented on packages/hardhat-polkadot/src/cli/port-project.ts:
```diff
@@ -0,0 +1,138 @@
+        throw new Error(`Something went wrong fetching ${MODULE} metadata from the npm registry`)
+    }
+
+    const polkadotHHVersion: string = m.version
+    const HHVersionRange = m.peerDependencies.hardhat
+    const minHHVersion = minVersion(HHVersionRange)!
+
+    // Ensure `hardhat` is within expected range
+    const HHLocation = pkg.dependencies?.hardhat ? "dependencies" : "devDependencies"
+    const currentHHVersion = coerce(pkg[HHLocation]?.hardhat ?? "0.0.0") ?? minVersion("0.0.0")!
+    pkg[HHLocation] ??= {}
+    if (lt(currentHHVersion, minHHVersion)) {
+        pkg[HHLocation].hardhat = `^${minHHVersion.version}`
+    } else if (gtr(currentHHVersion, HHVersionRange)) {
+        throw new Error(
+            `Unsupported hardhat version ${currentHHVersion.version}. Please manually install a compatible version (${HHVersionRange})`,
+        )
+    }
```
We're using the semver package— I haven't found any packages which handle this better but I can definitely keep looking into it for a potential refactor after I merge!
